README.md
AWS CloudWatch LOGGER [![Build Status](https://travis-ci.org/pjhl/aws-cwlogger.svg?branch=master)](https://travis-ci.org/pjhl/aws-cwlogger)
---------------------

This is simple logger to send logs into
[AWS CloudWatch Logs](http://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/WhatIsCloudWatchLogs.html) stream.

### Usage:

```javascript
const CWLogger = require('aws-cwlogger');
const logger = new CWLogger({
  // AWS access
  accessKeyId: "XXXXXXXXXXXXXXXXXXXX",
  secretAccessKey: "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
  region: "eu-central-1",
  // Logs destination
  groupName: "test_group",
  streamName: "test_stream"
});

logger.log('Text message');
logger.info({text: 'Supported all data types!'});
logger.warn(['Including'], {_: 'multiple arguments'});
logger.error(false);

logger.cat('category-name').log('Categories supported.');
logger.cat('category-name').info('with');
logger.cat('category-name').warn('all 4');
logger.cat('category-name').error('methods');
```

After that you should see in console:

```text
Text message
{ text: 'Supported all data types!' }
[ 'Including' ] { _: 'multiple arguments' }
false
<<category-name>>: Categories supported.
<<category-name>>: with
<<category-name>>: all 4
<<category-name>>: methods
```

And in **AWS console - CloudWatch - Logs**:

![AWS CloudWatch logs screen](/docs/screen-aws-cloudwatch-logs.png)

### Options:

| Param             | Type          | Description |
|-------------------|---------------|-------------|
| enabled           | {boolean}     | Enable sending logs to CloudWatch (default: `true`). Leave the option `false` to don't save logs into CloudWatch Logs |
| accessKeyId*      | {string}      | AWS API key (not required if `enabled === false`) |
| secretAccessKey*  | {string}      | AWS API secret (not required if `enabled === false`) |
| region*           | {string}      | AWS region (default: `eu-central-1`) (not required if `enabled === false`) |
| groupName*        | {string}      | CloudWatch log group name (not required if `enabled === false`) |
| streamName*       | {string}      | CloudWatch log stream name in group (not required if `enabled === false`) |
| consolePrint      | {boolean}     | Print errors into console (default: `true`) |
| flushDelay        | {number}      | Delay in ms. between sending messages (default: `1000`) |
| maxBatchCount     | {number}      | Max messages to send in butch (default: 5000) |
| maxBatchSize      | {number}      | The maximum batch size (default: 1000000) |

    P.S. IAM user should have permissions for actions
    ["logs:DescribeLogGroups","logs:DescribeLogStreams","logs:PutLogEvents"]
    P.P.S. Do not use root user for the security reasons!


INSTALL
-------

```bash
npm install aws-cwlogger --save
```


RUN TESTS
---------

### 1. Create CloudWatch Logs group and stream:

1.  Go to "AWS Console" - "Services" - "CloudWatch" - Tab "Logs"
2.  Create group (ex. `cwlogger-test`)
3.  Go to created group and create stream (ex. `test`)
4.  Go back to tab "Logs" and enabled column "ARN" in table.
5.  Copy your group's ARN in clipboard, you need it on the next step.

### 2. Create AWS IAM user (ex. `cwlogger-test`)

*   Make checked **Programmatic access**
*   Don't check **AWS Management Console access**

With permission:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "logs:DescribeLogGroups",
                "logs:DescribeLogStreams",
                "logs:PutLogEvents"
            ],
            "Resource": [
                "arn:aws:logs:eu-central-1:665544332211:log-group:cwlogger-test:*"
            ]
        }
    ]
}
```

**Don't forget to replace "Resource" with your ARN.**

### 3. Configure test

```bash
cp test/config/example.config.js test/config/config.js
nano test/config/config.js
```

### 4. Run tests

```bash
npm install
npm test
```

LICENSE
-------

MIT

CHANGELOG.md
CHANGELOG
=========


## v1.0.1 <small>\[2017-10-20\]</small>

*   Added option `options.enabled`
*   Fixed the orthographic mistakes in the README file


## v1.0.0 <small>\[2017-10-19\]</small>

Initial release.

test/index_test.js
const assert = require('assert');
const CWLogger = require('../libs/index');
// Local config for AWS
const config = require('./config/config');

describe("CWLogger", function () {

  this.timeout(10000);

  it('should push logs on CloudWatch Logs', function (done) {
    const logger = new CWLogger(Object.assign({}, config, {
      consolePrint: false,
      flushDelay: 10
    }));

    assert(logger instanceof CWLogger);

    logger.log('Who let the dogs out?');
    logger.info({text: 'Search for dogs', rewarding: '$20'});
    logger.warn(['Rewarding', 'is', 'a', 'lot']);
    logger.error({view: '#@)₴?$0'}, [5, '/', 5]);

    logger.cat('test-cat').log('Who let the dogs out?');
    logger.cat('test-cat').info({text: 'Search for dogs', rewarding: '$20'});
    logger.cat('test-cat').warn(['Rewarding', 'is', 'a', 'lot']);
    logger.cat('test-cat').error({view: '#@)₴?$0'}, [5, '/', 5]);

    logger.on('success', messagesSent => {
      assert.strictEqual(messagesSent.length, 8);
      assert.strictEqual(logger.count, 0, 'There is no messages in queue (all sent)');
      logger.destroy();
      done();
    });
  });

  it('should not send to CloudWatch', function () {
    const logger = new CWLogger({
      enabled: false, // We test this option
      consolePrint: true
    });

    logger.log('Test message');
    assert.strictEqual(logger.count, 0, 'The messages aren\'t saving');
    logger.destroy();
  });

});

test/Logger_test.js
const assert = require('assert');
const Logger = require('../libs/Logger');

describe('Logger', function () {

  describe('console-like methods', function () {

    it('should support simple (without category) run', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1});
      logger.onSend = (messages, resolve, reject) => {
        assert(Array.isArray(messages));
        assert.strictEqual(typeof resolve, 'function');
        assert.strictEqual(typeof reject, 'function');
        assert.strictEqual(messages.length, 4);
        assert.strictEqual(messages[0].message, '{"level":"log","msg":"qwerty"}');
        assert(!!messages[0].timestamp);
        assert.strictEqual(messages[1].message, '{"level":"info","msg":"qwerty"}');
        assert(!!messages[1].timestamp);
        assert.strictEqual(messages[2].message, '{"level":"warn","msg":"qwerty"}');
        assert(!!messages[2].timestamp);
        assert.strictEqual(messages[3].message, '{"level":"error","msg":"qwerty"}');
        assert(!!messages[3].timestamp);
        assert.strictEqual(logger.count, 4);
        resolve();
        assert.strictEqual(logger.count, 0);
        logger.destroy();
        done();
      };
      // All log methods
      logger.log('qwerty');
      assert.strictEqual(logger.count, 1);
      logger.info('qwerty');
      assert.strictEqual(logger.count, 2);
      logger.warn('qwerty');
      assert.strictEqual(logger.count, 3);
      logger.error('qwerty');
      assert.strictEqual(logger.count, 4);
    });

    it('should support categories', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1});
      logger.onSend = (messages, resolve, reject) => {
        assert(Array.isArray(messages));
        assert.strictEqual(typeof resolve, 'function');
        assert.strictEqual(typeof reject, 'function');
        assert.strictEqual(messages.length, 4);
        assert.strictEqual(messages[0].message, '{"level":"log","cat":"test","msg":"qwerty"}');
        assert(!!messages[0].timestamp);
        assert.strictEqual(messages[1].message, '{"level":"info","cat":"test","msg":"qwerty"}');
        assert(!!messages[1].timestamp);
        assert.strictEqual(messages[2].message, '{"level":"warn","cat":"test","msg":"qwerty"}');
        assert(!!messages[2].timestamp);
        assert.strictEqual(messages[3].message, '{"level":"error","cat":"test","msg":"qwerty"}');
        assert(!!messages[3].timestamp);
        assert.strictEqual(logger.count, 4);
        resolve();
        assert.strictEqual(logger.count, 0);
        logger.destroy();
        done();
      };
      // All log methods
      logger.cat('test').log('qwerty');
      assert.strictEqual(logger.count, 1);
      logger.cat('test').info('qwerty');
      assert.strictEqual(logger.count, 2);
      logger.cat('test').warn('qwerty');
      assert.strictEqual(logger.count, 3);
      logger.cat('test').error('qwerty');
      assert.strictEqual(logger.count, 4);
    });
  });

  describe('#onSend()', function () {
    it('should to send one message', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1});
      logger.onSend = (messages, resolve, reject) => {
        assert(Array.isArray(messages));
        assert.strictEqual(typeof resolve, 'function');
        assert.strictEqual(typeof reject, 'function');
        assert.strictEqual(messages.length, 1);
        assert.strictEqual(messages[0].message, '{"level":"info","msg":"qwerty"}');
        assert(!!messages[0].timestamp);
        assert.strictEqual(logger.count, 1);
        resolve();
        assert.strictEqual(logger.count, 0);
        logger.destroy();
        done();
      };
      logger.info('qwerty');
    });

    it('should to send multiple messages', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1, maxBatchCount: 2});
      let count = 0;
      logger.onSend = (messages, resolve, reject) => {
        // Method "onSend" must be executed twice due to batch size limit
        if (count++ === 0) {
          assert(Array.isArray(messages));
          assert.strictEqual(typeof resolve, 'function');
          assert.strictEqual(typeof reject, 'function');
          assert.strictEqual(messages.length, 2); // Batch size limit: 2
          assert.strictEqual(messages[0].message, '{"level":"log","msg":"qwerty"}');
          assert.strictEqual(messages[1].message, '{"level":"warn","msg":"asdfgh"}');
          assert.strictEqual(logger.count, 3);
          resolve();
          assert.strictEqual(logger.count, 1);
        } else {
          assert(Array.isArray(messages));
          assert.strictEqual(typeof resolve, 'function');
          assert.strictEqual(typeof reject, 'function');
          assert.strictEqual(messages.length, 1);
          assert.strictEqual(messages[0].message, '{"level":"log","msg":"zxcvbn"}');
          assert.strictEqual(logger.count, 1);
          resolve();
          assert.strictEqual(logger.count, 0);
          logger.destroy();
          done();
        }
      };
      logger.log('qwerty');
      logger.warn('asdfgh');
      logger.log('zxcvbn');
    });

    it('should support #reject()', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1});
      let count = 0;
      logger.onSend = (messages, resolve, reject) => {
        if (count++ === 0) {
          assert(Array.isArray(messages));
          assert.strictEqual(typeof resolve, 'function');
          assert.strictEqual(typeof reject, 'function');
          assert.strictEqual(messages.length, 1);
          assert.strictEqual(messages[0].message, '{"level":"info","msg":"qwerty"}');
          assert.strictEqual(logger.count, 1);
          reject();
          assert.strictEqual(logger.count, 1);
        } else {
          // This block should run on second try
          assert(Array.isArray(messages));
          assert.strictEqual(typeof resolve, 'function');
          assert.strictEqual(typeof reject, 'function');
          assert.strictEqual(messages.length, 1);
          assert.strictEqual(messages[0].message, '{"level":"info","msg":"qwerty"}');
          resolve();
          assert.strictEqual(logger.count, 0);
          logger.destroy();
          done();
        }
      };
      logger.info('qwerty');
    });

    it('should not allow #resolve() twice', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1});
      logger.onSend = (messages, resolve) => {
        resolve();
        assert.throws(resolve, Error);
        logger.destroy();
        done();
      };
      logger.info('qwerty');
    });

    it('should not allow #reject() twice', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1});
      logger.onSend = (messages, resolve, reject) => {
        reject();
        assert.throws(reject, Error);
        logger.destroy();
        done();
      };
      logger.info('qwerty');
    });

    it('should not allow #resolve() then #reject()', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1});
      logger.onSend = (messages, resolve, reject) => {
        resolve();
        assert.throws(reject, Error);
        logger.destroy();
        done();
      };
      logger.info('qwerty');
    });

    it('should not allow #reject() then #resolve()', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1});
      logger.onSend = (messages, resolve, reject) => {
        reject();
        assert.throws(resolve, Error);
        logger.destroy();
        done();
      };
      logger.info('qwerty');
    });

    it('should suport "maxBatchSize" limit', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1, maxBatchSize: 50});
      let count = 0;
      logger.onSend = (messages, resolve, reject) => {
        // Method "onSend" must be executed twice due to batch size limit
        if (count++ === 0) {
          assert(Array.isArray(messages));
          assert.strictEqual(typeof resolve, 'function');
          assert.strictEqual(typeof reject, 'function');
          assert.strictEqual(messages.length, 1);
          assert.strictEqual(messages[0].message, '{"level":"error","msg":"one"}');
          assert.strictEqual(logger.count, 2);
          resolve();
          assert.strictEqual(logger.count, 1);
        } else {
          assert(Array.isArray(messages));
          assert.strictEqual(typeof resolve, 'function');
          assert.strictEqual(typeof reject, 'function');
          assert.strictEqual(messages.length, 1);
          assert.strictEqual(messages[0].message, '{"level":"error","msg":"two"}');
          assert.strictEqual(logger.count, 1);
          resolve();
          assert.strictEqual(logger.count, 0);
          logger.destroy();
          done();
        }
      };
      logger.error('one');
      logger.error('two');
    });

    it('should support different log types and arguments count', function (done) {
      const logger = new Logger({consolePrint: false, flushDelay: 1});
      logger.onSend = (messages, resolve, reject) => {
        assert(Array.isArray(messages));
        assert.strictEqual(typeof resolve, 'function');
        assert.strictEqual(typeof reject, 'function');
        assert.strictEqual(messages.length, 7);
        assert.strictEqual(messages[0].message, '{"level":"log","msg":"true"}');
        assert.strictEqual(messages[1].message, '{"level":"log","msg":"{ a: 1 }"}');
        assert.strictEqual(messages[2].message, '{"level":"log","msg":"[]"}');
        assert.strictEqual(messages[3].message, '{"level":"log","msg":"[ 1, 2, 3 ]"}');
        assert.strictEqual(messages[4].message, '{"level":"log","msg":"NaN"}');
        assert.strictEqual(messages[5].message, '{"level":"warn","msg":"1 2"}');
        assert.strictEqual(messages[6].message, '{"level":"warn","msg":"{ a: 1 } [ 2, 3 ]"}');
        assert.strictEqual(logger.count, 7);
        resolve();
        assert.strictEqual(logger.count, 0);
        logger.destroy();
        done();
      };
      // Different args
      logger.log(true);
      logger.log({a: 1});
      logger.log([]);
      logger.log([1, 2, 3]);
      logger.log(NaN);
      // More then 1 argument
      logger.warn(1, 2);
      logger.warn({a: 1}, [2, 3]);
    });

  });
});

test/CloudWatchLogs_test.js
const assert = require('assert');
const CloudWatchLogs = require('../libs/CloudWatchLogs');
// Local config for AWS
const config = require('./config/config');

describe("CloudWatchLogs", function () {

  this.timeout(10000);

  it('should post messages', function () {
    return new CloudWatchLogs(config)
      .put([{
        message: "Test_message1",
        timestamp: new Date().getTime()
      }])
      .then(() => {
        assert(true);
      });
  });

  it('should not post messages in nonexistent group', function (done) {
    new CloudWatchLogs(Object.assign({}, config, {groupName: 'UN' + Math.random()}))
      .put([{
        message: "Test_message2",
        timestamp: new Date().getTime()
      }])
      .then(() => {
        done('Must be an unreachable');
      })
      .catch(err => {
        assert.strictEqual(err.code, 'AccessDeniedException');
        done();
      });
  });

  it('should not post messages in nonexistent stream', function (done) {
    new CloudWatchLogs(Object.assign({}, config, {streamName: 'UN' + Math.random()}))
      .put([{
        message: "Test_message3",
        timestamp: new Date().getTime()
      }])
      .then(() => {
        done('Must be an unreachable');
      })
      .catch(err => {
        assert(err.message.indexOf('is not found in group') !== -1);
        done();
      });
  });

});

libs/CloudWatchLogs.js
const AWS = require('aws-sdk');


/**
 * CloudWatch Logs integration.
 * Usage example:
 *
 * new CloudWatchLogs({
 *   // AWS access
 *   accessKeyId: "XXXXXXXXXXXXXXXXXXXX",
 *   secretAccessKey: "XXXXXXxXXXXXXXXxXXxXXxXXXXXXXXXxXXXXXxXX",
 *   region: "eu-central-1",
 *   // Logs destination
 *   groupName: "test_group",
 *   streamName: "test_stream"
 * })
 * .put([{
 *    message: "Message body",
 *    timestamp: new Date().getTime()
 *  }])
 * .then(() => {
 *    // Put success
 *  })
 * .catch(err => {
 *    // Process error
 *    throw err;
 *  });
 */
class CloudWatchLogs {

  constructor(options) {
    const {
      enabled = true,
      accessKeyId = null,
      secretAccessKey = null,
      region = 'eu-central-1',
      groupName = '',
      streamName = ''
    } = options;

    this.enabled = enabled;
    this.sequenceToken = null;
    this.groupName = groupName;
    this.streamName = streamName;

    if (this.enabled) {
      if (!accessKeyId) {
        throw new Error('CloudWatchLogs constructor: option "accessKeyId" is required.');
      }
      if (!secretAccessKey) {
        throw new Error('CloudWatchLogs constructor: option "secretAccessKey" is required.');
      }
      if (!groupName) {
        throw new Error('CloudWatchLogs constructor: option "groupName" is required.');
      }
      if (!streamName) {
        throw new Error('CloudWatchLogs constructor: option "groupName" is required.');
      }

      // Create Amazon api object
      this._awsCWL = new AWS.CloudWatchLogs({
        apiVersion: '2014-03-28',
        accessKeyId: accessKeyId,
        secretAccessKey: secretAccessKey,
        region: region
      });
    }
  }

  /**
   * This function updates "sequenceToken" from existent stream (this.sequenceToken)
   * @return {Promise}
   * @private
   */
  _getToken() {
    return new Promise((resolve, reject) => {
      this._awsCWL
        .describeLogStreams({
          logGroupName: this.groupName,
          logStreamNamePrefix: this.streamName
        }, (err, data) => {
          if (err) {
            return reject(err);
          }
          const stream = data.logStreams.find(val => val.logStreamName === this.streamName);
          if (!stream) {
            return reject(new Error(`CloudWatchLogs error: stream "${this.streamName}" is not found in group "${this.groupName}"`));
          }
          this.sequenceToken = stream.uploadSequenceToken;
          resolve();
        });
    });
  }

  /**
   * Put log events on AWS CloudWatch logs
   * @param {[]} messages   List of messages: {message: {string}, timestamp: {number}} (see new Date().getTime())
   * @return {Promise}
   */
  put(messages) {
    if (!this.enabled) {
      return new Promise((resolve, reject) => reject('CloudWatch option "enabled": false'));
    }
    if (!this.sequenceToken) {
      // Get token then put
      return this
        ._getToken()
        .then(() => this._put(messages));
    }
    return this._put(messages);
  }

  _put(messages) {
    return new Promise((resolve, reject) => {
      if (!Array.isArray(messages)) {
        return reject('Param "messages" must be an array.');
      }

      this._awsCWL.putLogEvents({
        logGroupName: this.groupName,
        logStreamName: this.streamName,
        logEvents: messages,
        sequenceToken: this.sequenceToken
      }, (err, data) => {
        if (err) {
          if (err.code === 'InvalidSequenceTokenException') {
            // Repeat after getting token
            return resolve(this._getToken().then(() => this._put(messages)));
          }
          return reject(err);
        }
        this.sequenceToken = data.nextSequenceToken;
        resolve();
      });
    });
  }
}

module.exports = CloudWatchLogs;

libs/Logger.js
const util = require('util');
const EventEmitter = require('events');

const OPTIONS = Symbol();
const QUEUE = Symbol();

class Logger extends EventEmitter {

  /**
   * Create Logger
   * @param {object} options    Options
   */
  constructor(options = {}) {
    super();
    const {enabled = true, consolePrint = true, flushDelay = 1000, maxBatchCount = 5000, maxBatchSize = 1000000} = options;
    // Options
    this[OPTIONS] = {
      /**
       * Enable saving logs to queue (default: true)
       * @type {boolean}
       */
      enabled: enabled,
      /**
       * Print errors into console (default: true)
       * @type {boolean}
       */
      consolePrint: consolePrint,
      /**
       * Delay in ms. between sending messages (default: 1000)
       * @type {number}
       */
      flushDelay: flushDelay,
      /**
       * Max messages to send in butch (default: 5000).
       * The real maximum number of log events in a batch is 10,000.
       * @type {number}
       */
      maxBatchCount: maxBatchCount,
      /**
       * The maximum batch size (default: 1000000).
       * The real max size is 1,048,576 bytes, and this size is calculated as the sum of all event messages in UTF-8,
       * plus 26 bytes for each log event.
       * @type {number}
       */
      maxBatchSize: maxBatchSize
    };
    // Queue of messages: {message: {string}, timestamp: {number}}
    this[QUEUE] = [];

    // Start timer to send log events periodically
    this._startTimer();
  }

  /**
   * Log
   * @param level
   * @param args
   * @private
   */
  _log(level = 'log', ...args) {
    // With category
    let cat = undefined, msg = '';
    if (args.length === 1 && typeof args[0] === 'object' && args[0].$category) {
      cat = args[0].$category;
      msg = util.format(...args[0].$messages);
      // Print into console
      if (this[OPTIONS].consolePrint) {
        console[level](`<<${cat}>>:`, ...args[0].$messages);
      }
    } else {
      msg = util.format(...args);
      // Print into console
      if (this[OPTIONS].consolePrint) {
        console[level](...args);
      }
    }
    if (this[OPTIONS].enabled) {
      // Add to queue
      this[QUEUE].push({
        message: JSON.stringify({
          level: level,
          cat: cat,
          msg: msg
        }),
        timestamp: new Date().getTime()
      });
    }
  }

  log(...args) {
    return this._log('log', ...args);
  }

  info(...args) {
    return this._log('info', ...args);
  }

  warn(...args) {
    return this._log('warn', ...args);
  }

  error(...args) {
    return this._log('error', ...args);
  }

  /**
   * Set log category
   * @param {string} name
   * @return {object}
   */
  cat(name = '') {
    const $this = this;
    return {
      log(...args) {
        return $this.log({$category: name, $messages: args});
      },
      info(...args) {
        return $this.info({$category: name, $messages: args});
      },
      warn(...args) {
        return $this.warn({$category: name, $messages: args});
      },
      error(...args) {
        return $this.error({$category: name, $messages: args});
      }
    }
  }

  /**
   * Start timer to send log events periodically
   * @private
   */
  _startTimer() {
    if (this._destroyed || this._timer) {
      return;
    }
    this._timer = setTimeout(() => {
      this._timer = null;
      this._sendChunk();
    }, this[OPTIONS].flushDelay);
  }

  /**
   * Send chunk of messages
   * @private
   */
  _sendChunk() {
    if (this[QUEUE].length > 0) {
      // Search messages
      let batchSize = 0, posEnd = 0;
      for (let i = 0; i < this[QUEUE].length; i++) {
        const msg = this[QUEUE][i].message;
        // Current message size (+ 26 bytes overhead)
        const currentMessageSize = msg.length + 26;
        // Check limit "maxBatchSize"
        if (batchSize + currentMessageSize > this[OPTIONS].maxBatchSize) {
          break;
        }
        // Check limit "maxBatchCount"
        if (i >= this[OPTIONS].maxBatchCount) {
          break;
        }
        posEnd = i;
        batchSize += currentMessageSize;
      }
      // Slice messages
      const messages = this[QUEUE].slice(0, posEnd + 1);
      if (messages.length === 0) {
        this._startTimer();
        return;
      }
      // Create functions
      let executed = '';
      const resolve = () => {
        if (executed !== '') {
          throw new Error(`Method "${executed}" was executed before`);
        }
        executed = 'resolve';
        this[QUEUE].splice(0, posEnd + 1);
        this._startTimer();
      };
      const reject = () => {
        if (executed !== '') {
          throw new Error(`Method "${executed}" was executed before`);
        }
        executed = 'reject';
        this._startTimer();
      };

      this.onSend(messages, resolve, reject);
      return;
    }
    this._startTimer();
  }

  /**
   * Send batch event
   * @param {[]} messages       List of messages to send: {message: {string}, timestamp: {number}}
   * @param {function} resolve  Approve sending
   * @param {function} reject   Reject sending (it will be repeated after a while)
   */
  onSend(messages, resolve, reject) {
    this.destroy();
    throw new Error('Function "onSend" must be overridden.');
  }

  /**
   * Destroy logger (shutdown timer and stop sending messages)
   */
  destroy() {
    this._destroyed = true;
    if (this._timer) {
      clearTimeout(this._timer);
    }
  }

  /**
   * Returns count of elements in queue
   */
  get count() {
    return this[QUEUE].length;
  }

}

module.exports = Logger;